### PR TITLE
Feature/#3 sdk component manager

### LINKED ISSUE #3: SDK Component manager
https://github.com/smartknob-ha/firmware-idf/issues/11

Add a component that serves as a thread to run other components in, with partially automatic error handling.

### PR DESCRIPTION
**Pull request by SquaredPotato**
Closes #3 , coupled to https://github.com/smartknob-ha/firmware-idf/pull/15

This adds a manager that can be used to handle component type objects to be run in a single thread, by implementing a simple interface.

### CODE REVIEW

File: manager/Kconfig [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+    menu "Manager"
+
+        config NUM_COMPONENTS
+            int "Number of components in use from the SDK"
+            default 1
+            help
+                You should probably not touch this unless you've added a component
+    endmenu
+endmenu
```
Comment (abrondijk) [inline manager/Kconfig:11]: There is a Kconfig entry for `NUM_COMPONENTS`, but it is not used in any of the sourcefiles and is replaced with a define of the same name.

File: manager/include/component.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+         * @note  Any error message returned will be treated
+         *          as an error. The manager will try to restart
+         *          the component by calling stop(), initialize()
+         *          and run().
+        */
+        virtual res run() { return Ok(ESP_OK);  };
+
+        /**
+         * @brief May be called after the component has returned
+         *        an error status, or when it's shutting down
```
Comment (abrondijk) [inline manager/include/component.hpp]: Should be worded better, as it's used to shutdown components.

**Review by abrondijk — CHANGES_REQUESTED** [review]

File: manager/include/component.hpp (hunk 2 above)
Comment (SquaredPotato) [reply manager/include/component.hpp]: Like this?

Comment (abrondijk) [reply manager/include/component.hpp]: _yes_

**Review by abrondijk — APPROVED** [review]

**Pull request merged**